stream-rpc/src/connection.rs

```diff
@@ -0,0 +1,389 @@
+            ])
+            .inc();
+        self.send_error(id, error).await
+    }
+
+    pub fn unsubscribe(&self, id: &RequestIdentifier) -> Option<()> {
+        self.running_tasks.stop(id)
+    }
+
+    pub fn is_closed(&self) -> bool {
+        return self.sender.is_closed();
+    }
+
+    pub async fn received_message(&self, db: Arc<dyn DbReader>, message: String) {
+        match JsonRpcRequest::from_str(&message) {
+            Ok(request) => {
+                debug!(
+                    logging::SubscriptionRequestLog {
+                        transport: self.connection_context.transport.as_str(),
+                        remote_addr: self.connection_context.remote_addr.clone(),
+                        client_id: self.id,
+                        rpc_method: Some(request.method_name()),
+                    },
+                    "subscription request"
+                );
+                if let Err(err) = self.handle_rpc_request(db, &request) {
+                    self.send_error_with_metrics(
+                        Some(request.method_request.method()),
+                        Some(request.id),
+                        Some(err),
+                    )
+                        .await
+                        .ok();
+                }
+            }
+            Err((err, method, id)) => {
+                // We couldn't parse the request- it's not valid json or an unknown structure
+                debug!(
+                    logging::SubscriptionRequestLog {
+                        transport: self.connection_context.transport.as_str(),
+                        remote_addr: self.connection_context.remote_addr.clone(),
+                        client_id: self.id,
+                        rpc_method: method.map_or(None, |v| Some(v.as_str())),
+                    },
+                    "failed to parse subscription request ({})", &err
+                );
+                metric_subscription_rpc_received(
+                    &self,
+                    method.map_or("unknown", |m| m.as_str()),
+                    counters::RpcResult::Error,
+                );
+                self.send_error_with_metrics(method, id, Some(err))
+                    .await
+                    .ok();
+            }
+        }
+    }
+
+    fn handle_rpc_request(
+        &self,
+        db: Arc<dyn DbReader>,
+        request: &JsonRpcRequest,
+    ) -> Result<(), JsonRpcError> {
+        let mut tasks = self.running_tasks.tasks.lock().unwrap();
+        if tasks.contains_key(&request.id) {
+            debug!(
+                "Client#{} already has a subscription for '{}'",
+                self.id, &request.id
+            );
+            let err = JsonRpcError::invalid_request_with_msg(format!(
+                "Subscription for '{}' already exists",
+                &request.id
+            ));
+
+            return Err(err);
+        }
+
+        match request.call_method(db, self.clone()) {
```

Comment:
is clone right? or should pass in the reference

Reply:
it can't pass a reference because the `client` has to be `moved` across threads in the async closure for the tokio task. Does that make sense?